### PR TITLE
[REF] Cleanup function for retrieving contact types.

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
Minor code cleanup intended to allow us to deprecate a couple of odd params in ```CRM_Core_DAO::executeQuery``` - naming $abort = FALSE & $daoName.

Before
----------------------------------------
Use of older methods

After
----------------------------------------
Retrieves using apiv4 - DB retrieves & caches once & filters as needed - should be so cheap as to not warrant caching the filtering

Technical Details
----------------------------------------
This switches to using the apiv4 as part of trying to eliminate calls to executeQuery where the DAO is passed in
like

```
       = CRM_Core_DAO::executeQuery(, [],
        FALSE, 'CRM_Contact_DAO_ContactType'
      );
```

Note that this file seems to be the only place this is done.

Also note that I added tests first & the changes to the test highlight what is changed in the PR in the output - namely
- stricter type casting (courtesy apiv4)
- same keys present for all types

Comments
----------------------------------------

### CODE REVIEW

**Comment by seamuslee001** [comment]
Jenkins re test this please

File: CRM/Contact/BAO/ContactType.php [hunk 1]
```diff
@@ -886,11 +869,37 @@ public static function deleteCustomRowsOfSubtype($gID, $subtypes = [], $subtypes
...
 
+  /**
+   * Get all contact types, leveraging caching.
+   *
+   * @return array
+   *
+   * @throws \API_Exception
+   */
+  protected static function getAllContactTypes() {
+    if (!Civi::cache('contactTypes')->has('all')) {
```
Comment (seamuslee001) [inline CRM/Contact/BAO/ContactType.php:889]: @eileenmcnaughton do we need to bust this cache whenever a contact type is enabled or disabled or deleted or does the system already do that sensibly?

Comment (eileenmcnaughton) [reply CRM/Contact/BAO/ContactType.php:889]: It's already doing ```    Civi::cache('contactTypes')->clear();```

**Comment by seamuslee001** [comment]
This seems fine to me and unit tests should cover us here I think

**Pull request merged**